### PR TITLE
BUG: Fix problem with SparseDataFrame not persisting to csv

### LINKED ISSUE #19384: SparseDataFrame to_csv returns IndexError: too many indices for array
Calling to_csv on a SparseDataFrame returns IndexError: too many indices for array.
```python
import pandas as pd
from scipy.sparse import rand

x = rand(100, 10001, density=0.2, format='csr')
df = pd.SparseDataFrame(x)
df.to_csv("tmp.csv")
```

Calling .to_dense() on the df prior to writing out, of course, works fine.
```python
df.to_dense().to_csv("tmp.csv")
```

I'd like to be able to save sparse data to disk.  If for some reason it shouldn't be written out this way, is there any way to add that in the doc or the error message?

#### Expected Output

Nontype, no exception, and a csv on the file system.

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Darwin
OS-release: 16.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.21.0
pytest: None
pip: 9.0.1
setuptools: 28.8.0
Cython: None
numpy: 1.13.3
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.1.0
psycopg2: 2.7.3.2 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by hexgnu**
- [x] closes #19384
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/internals.py [hunk 1]
```diff
@@ -704,7 +704,9 @@ def to_native_types(self, slicer=None, na_rep='nan', quoting=None,
                         **kwargs):
         """ convert to our native types format, slicing if desired """
 
-        values = self.values
+        # values = self.values
```
Comment (jreback) [inline pandas/core/internals.py]: you can remove the comment

File: pandas/tests/io/formats/test_to_csv.py [hunk 2]
```diff
@@ -253,3 +253,12 @@ def test_to_csv_string_array_utf8(self):
             df.to_csv(path, encoding='utf-8')
             with open(path, 'r') as f:
                 assert f.read() == expected_utf8
+
+    def test_to_csv_sparse_dataframe(self):
+        sdf = pd.SparseDataFrame({'a': [1, 2]})
```
Comment (jreback) [inline pandas/tests/io/formats/test_to_csv.py]: add a comment here for the issue number. I would rather have this test in pandas/tests/sparse/frame/test_to_csv.py 

can you add several variations, e.g. using fill value and with some nulls.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by hexgnu** [comment]
K updated thanks @jreback

**Comment by TomAugspurger** [comment]
@hexgnu do the tests you've written pass for you locally?

**Comment by hexgnu** [comment]
oi sorry, should pass now.

**Comment by TomAugspurger** [comment]
Still one lint failure: https://travis-ci.org/pandas-dev/pandas/jobs/335405859#L3035

I'm looking into the (unrelated) S3 failures later this morning.

**Comment by hexgnu** [comment]
Ok for real this time it passed @jreback @TomAugspurger thanks 😄

File: pandas/tests/sparse/frame/test_to_csv.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import numpy as np
+from pandas import SparseDataFrame, read_csv
+from pandas.util import testing as tm
+
+
+def test_to_csv_sparse_dataframe():
+    fill_values = [np.nan, 0, None, 1]
+
+    for fill_value in fill_values:
```
Comment (jreback) [inline pandas/tests/sparse/frame/test_to_csv.py]: can you parameterize

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/sparse/frame/test_to_csv.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import numpy as np
+from pandas import SparseDataFrame, read_csv
+from pandas.util import testing as tm
+
+
+def test_to_csv_sparse_dataframe():
+    fill_values = [np.nan, 0, None, 1]
```
Comment (jreback) [inline pandas/tests/sparse/frame/test_to_csv.py]: can you add the issue number here as well

File: doc/source/whatsnew/v0.23.0.txt [hunk 5]
```diff
@@ -497,7 +497,7 @@ I/O
 - Bug in :func:`DataFrame.to_parquet` where an exception was raised if the write destination is S3 (:issue:`19134`)
 - :class:`Interval` now supported in :func:`DataFrame.to_excel` for all Excel file types (:issue:`19242`)
 - :class:`Timedelta` now supported in :func:`DataFrame.to_excel` for xls file type (:issue:`19242`, :issue:`9155`)
--
+- Bug in :class:`SparseDataFrame.to_csv` where too many indices for values (:issue:`19384`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: move to sparse, and don't need this kind of detail, just say was raising.

**Comment by jreback** [comment]
a lint error.ping when pushed and green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
great. ping on green.

**Comment by TomAugspurger** [comment]
Thanks @hexgnu!

**Pull request merged**